aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1092,6 +1092,77 @@ Tensor permute(const Tensor& self, IntArrayRef dims) {
   return self.as_strided(newSizes, newStrides);
 }
 
+Tensor permute_sparse_coo(const Tensor& self, IntArrayRef dims) {
+  const auto ndim = self.dim();
+  TORCH_CHECK(ndim == static_cast<int64_t>(dims.size()),
+      "permute(sparse_coo): number of dimensions in the tensor input ",
+      "does not match the length of the desired ordering of dimensions ",
+      "i.e. input.dim() = ", ndim, " is not equal to len(dims) = ", dims.size());
+
+  const auto sparse_ndim = self.sparse_dim();
+  const auto dense_ndim = self.dense_dim();
+
+  const auto old_sizes = self.sizes();
+  auto new_sizes = old_sizes.vec();
+  auto wrapped_dims_vec = dims.vec();
+  auto seen_dims = std::vector<bool>(ndim);
+  for (const auto i : c10::irange(ndim)) {
+    const auto d = maybe_wrap_dim(dims[i], ndim);
+    TORCH_CHECK(!seen_dims[d],
+        "permute(sparse_coo): duplicate dims are not allowed.");
+    seen_dims[d] = true;
+    wrapped_dims_vec[i] = d;
+    new_sizes[i] = old_sizes[d];
+  }
+  const auto wrapped_dims = IntArrayRef(wrapped_dims_vec);
```

Comment:
I don't think you're necessarily saving a lot here by using `IntArrayRef`. Also the call to `at::arange` is likely very slow (in terms of framework overhead, since it's creating an at::Tensor). I'd try implementing this small vector wrangling with plain std::vectors first and then we can employ `IntArrayRef` and `DimVector` as optimization. Maybe that'll uncover a bug within this function that resolves the autograd issue below.

Reply:
`IntArrayRef` provides a nice functionality with `slice`, so it is not about optimization in this context.

Reply:
I think this could the source of a hidden bug... `IntArrayRef` is not owning, right? It could be that `wrapped_dims_vecs` could get destroyed before `wrapped_dims` is accessed I suppose.